### PR TITLE
Allow Command to take more builder options

### PR DESCRIPTION
**Pull request by dragon512**
This adds the ability for Command to act more like a one-off builder. This is useful for cases in which other build systems, and allowing Scons to scanners to correctly understand the output for better implicit depends on handling

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by bdbaddog** [comment]
Can you edit the pull request template to include any useful info.. ;)
Also is it safe to say this is a work in progress? If so please change the Title of this pull request to indicate such [WIP] should be sufficient prefixed on the current title.

**Comment by dragon512** [comment]
@bdbaddog Not sure what is the best way to add 

ProgramScanner = SCons.Scanner.Prog.ProgramScanner()

I was going to add this Defaults.py, but the comments looks like we might have a more official way to get at the scanner??

Just wanted to check

**Comment by bdbaddog** [comment]
Can you add some updates to the docs to include this change?

File: src/engine/SCons/Environment.xml [hunk 1]
```diff
@@ -937,19 +937,28 @@ for a single special-case build.
...
+<varname>source_scanner</varname>,
+<varname>target_scanner</varname>,
+<varname>source_factory</varname>, and
+<varname>target_factory</varname>
+keyword arguments. This can
 be used to specify
 a Scanner object
-that will be used to scan the sources.
-(The global
+that will be used to scan node
```
Comment (bdbaddog) [inline src/engine/SCons/Environment.xml]: Why change from sources to node? That's less easily understandable.
Liekly you want separate paragraph for *_scanner than for *_factory. as they do different things.

Comment (dragon512) [reply src/engine/SCons/Environment.xml]: Let me make a new pass on the text

Comment (dragon512) [reply src/engine/SCons/Environment.xml]: @bdbaddog I updated the docs.. should be better than before

**Comment by dragon512** [comment]
@bdbaddog Just pinging again to see what else you need to be updated

File: src/engine/SCons/Environment.xml [hunk 2]
```diff
@@ -937,19 +937,29 @@ for a single special-case build.
...
+For example, the global
 <literal>DirScanner</literal>
 object can be used
 if any of the sources will be directories
 that must be scanned on-disk for
 changes to files that aren't
-already specified in other Builder of function calls.)
+already specified in other Builder of function calls.
+The *_factory args take a factory function that the
+Command will use to turn any inputs
```
Comment (bdbaddog) [inline src/engine/SCons/Environment.xml]: change to sources or targets from inputs.

**Comment by dragon512** [comment]
@bdbaddog I added the change for inputs to sources and targets

File: src/engine/SCons/Environment.xml (hunk 2 above)
Comment (dragon512) [reply src/engine/SCons/Environment.xml]: Fixed!

**Comment by bdbaddog** [comment]
Test failing under py27.   Please fix.
We've not yet dropped py 27 support.

```
209/1180 (17.71%) /home/travis/virtualenv/python2.7.14/bin/python -tt test/Command.py

/home/travis/build/SCons/scons/src/script/scons.py returned 2

STDOUT =========================================================================

scons: Reading SConscript files ...

STDERR =========================================================================

  File "/tmp/testcmd.3860.MxS7Wp/SConstruct", line 23

    print("Source scanner node=", node, "builder =", builder,file=sys.stderr)

                                                                 ^

SyntaxError: invalid syntax
```

**Comment by bdbaddog** [comment]
@mwichmann  & @garyo  - any concerns/comments? Assuming the tests pass I'm inclined to merge.

**Comment by mwichmann** [comment]
looks ok to me.

File: src/CHANGES.txt [hunk 3]
```diff
@@ -5,6 +5,9 @@
                             Change Log
 
 RELEASE  VERSION/DATE TO BE FILLED IN LATER
+  From Jason Kenny
+    - Update Command() function to accept target_scanner, source_factory, and target_factory arguments.
+      This make Command act more like a one-off builder.
```
Comment (garyo) [inline src/CHANGES.txt:10]: "This *makes*..."

File: src/engine/SCons/Environment.py [hunk 4]
```diff
@@ -1984,9 +1984,22 @@ def Command(self, target, source, action, **kw):
             'target_factory' : self.fs.Entry,
             'source_factory' : self.fs.Entry,
         }
+        # source scanner
         try: bkw['source_scanner'] = kw['source_scanner']
         except KeyError: pass
         else: del kw['source_scanner']
+        #target scanner
```
Comment (garyo) [inline src/engine/SCons/Environment.py:1991]: comments should have a space after the "#" (same below)

File: test/Command.py [hunk 5]
```diff
@@ -62,6 +63,19 @@ def sub(env, target, source):
                 t.write(s.read())
     return 0
 
+def source_scanner(node, env, path, builder):
+    print("Source scanner node=", node, "builder =", builder,file=sys.stderr)
+    return []
+
```
Comment (garyo) [inline test/Command.py:69]: extra blank line here (minor)

**Review by garyo — COMMENTED** [review]
Looks good to me! A few minor formatting tweaks would make it ideal.

**Comment by bdbaddog** [comment]
@garyo - fixed your comments in Master and just pushed.  I forgot to reload the PR and didn't see your comments until after I merged.

**Pull request merged**